### PR TITLE
Add support for excluding tests with TestFilter

### PR DESCRIPTION
**Pull request by felipecsl**
### Context
This change adds the ability to exclude the execution of individual test methods with the `--excludeTests` command line option for the `Test` task.

This has been previously discussed/requested [here](https://discuss.gradle.org/t/how-to-exclude-a-single-test-method/20860) and aims to solve a real use case we currently have.

This can be used in conjunction with `--tests` but using both options together may lead to confusing/unexpected results and should be discouraged IMHO.

Still TODO adding unit tests for this functionality, but wanted to open the PR to check if this is something that you folks would be open to merge.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by oehme** [comment]
What's the use case that this solves?

**Comment by felipecsl** [comment]
@oehme when running tests on CI, it'd be useful to run the entire suite but ignore a few individual tests, for example tests that we know are flaky/unstable.

**Comment by oehme** [comment]
I'm not convinced this is a common enough requirement to get a first-class option. Have you considered using `-PflakyTests=...` and passing that into `Test.exclude`?

**Comment by felipecsl** [comment]
Yes we have, but `Test.exclude` can only exclude entire test classes and not individual test methods, which is what we need. I'd hate to have maintain a forked build of Gradle indefinitely, how can I convince you otherwise? 😄

**Comment by felipecsl** [comment]
@oehme that could work! yeah we don't really need a dedicated command line switch for this, so a project property would totally work. I'll look into that and will update this PR accordingly. Thanks for your flexibility!

**Comment by felipecsl** [comment]
@oehme Alright, PR updated. This no longer has a dedicated command line switch and has instead been added directly to the `TestFilter`, as requested. The API is symmetric to `includeTest`/`includePatterns`.
I'm happy to add some integration tests if you have any pointers on the best approach.

**Comment by felipecsl** [comment]
friendly ping @oehme

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/DefaultTestFilter.java [hunk 1]
```diff
@@ -15,19 +15,20 @@
...
 
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.testing.TestFilter;
+
 public class DefaultTestFilter implements TestFilter {
 
-    private Set<String> testNames = new HashSet<String>();
-    private Set<String> commandLineTestNames = new HashSet<String>();
+    private Set<String> includeTestNames = new HashSet<String>();
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/DefaultTestFilter.java]: These can all be final.

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java [hunk 2]
```diff
@@ -56,35 +64,77 @@ public TestSelectionMatcher(Collection<String> includedTests, Collection<String>
...
+    private boolean isExactClassNameMatch(List<TestPattern> patterns, String fullQualifiedName) {
+        for (TestPattern pattern : patterns) {
+            if (pattern.matchesClass(fullQualifiedName)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean isClassNameMatch(List<TestPattern> patterns, String fullQualifiedName) {
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java]: This should be named `mayMatchClass`

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java [hunk 3]
```diff
@@ -56,35 +64,77 @@ public TestSelectionMatcher(Collection<String> includedTests, Collection<String>
...
+    }
+
+    private boolean mayExcludeClass(String fullQualifiedName) {
+        if (buildScriptExcludePatterns.isEmpty()) {
+            return false;
+        }
+        return isExactClassNameMatch(buildScriptExcludePatterns, fullQualifiedName);
+    }
+
+    private boolean isExactClassNameMatch(List<TestPattern> patterns, String fullQualifiedName) {
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java]: this should be named `matchesClass`

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java [hunk 4]
```diff
@@ -56,35 +64,77 @@ public TestSelectionMatcher(Collection<String> includedTests, Collection<String>
...
+            // When there is a class name match, return true for excluding it so that we can keep
+            // searching in individual test methods for an exact match. If we return false here
+            // instead, then we'll give up searching individual test methods and just ignore the
+            // entire test class, which may not be what we want.
+            return true;
+        }
+        return matchesIncludeOrExcludePattern(buildScriptExcludePatterns, className, methodName);
+    }
+
+    private boolean matchesIncludeOrExcludePattern(List<TestPattern> patterns, String className,
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java]: This should be named "matches class and method"

File: subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java [hunk 5]
```diff
@@ -56,35 +64,77 @@ public TestSelectionMatcher(Collection<String> includedTests, Collection<String>
...
-        for (TestPattern pattern : includePatterns) {
-            if (pattern.matchesClassAndMethod(fullQualifiedName, methodName)) {
+        return matchesIncludeOrExcludePattern(includePatterns, className, methodName);
+    }
+
+    private boolean matchesExcludePattern(String className, String methodName) {
+        if (buildScriptExcludePatterns.isEmpty()) {
+            return false;
+        }
+        if (isClassNameMatch(buildScriptExcludePatterns, className) && methodName == null) {
```
Comment (oehme) [inline subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java]: I don't understand this part and why it is different from the include code path. Can you please explain with an example?

Comment (oehme) [reply subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java]: @felipecsl Could you give me some feedback on this one please?

**Comment by oehme** [comment]
I will be on vacation for 2 weeks, but will come back to this before 5.0 goes out.

**Comment by felipecsl** [comment]
Thanks Stefan! Still pending for me to add integration tests to `JUnitFilteringIntegrationTest` and `TestNGFilteringIntegrationTest` (will try to do later today), but otherwise all other comments should be already addressed.

**Comment by oehme** [comment]
@felipecsl Can you rebase your branch on `release` please? As soon as that's done I'll run the tests and merge if they are green.

**Comment by felipecsl** [comment]
@oehme I didn't realize you had changed the base branch for this PR and I rebased against master instead. How should we proceed? It's basically just f23229a and 9fccc56

**Comment by oehme** [comment]
Please have a look at the [failed tests](https://builds.gradle.org/viewQueued.html?itemId=16747927).

**Comment by felipecsl** [comment]
`JUnitTestClassProcessorTest` tests fixed

**Comment by oehme** [comment]
The new methods are missing `@Incubating` and `@since 5.0`

**Comment by felipecsl** [comment]
New public API methods in `TestFilter` annotated

**Comment by oehme** [comment]
Thank you Felipe!

**Comment by felipecsl** [comment]
Awesome, thank YOU Stefan!

**Pull request merged**